src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public class JsonContent : HttpContent
+    {
+        internal const string JsonMediaType = "application/json";
+
+        private readonly byte[] _content;
+        private readonly int _offset;
+        private readonly int _count;
+
+        private static MediaTypeHeaderValue CreateMediaType(string mediaTypeAsString)
+        {
+            MediaTypeHeaderValue mediaType = MediaTypeHeaderValue.Parse(mediaTypeAsString);
+
+            // If the instantiated mediaType does not specify its CharSet, set UTF-8 by default.
+            if (mediaType.CharSet == null)
+            {
+                mediaType.CharSet = Encoding.UTF8.WebName;
+            }
+
+            return mediaType;
+        }
+
+        public Type ObjectType { get; }
+
+        public object? Value { get; }
+
+        public JsonContent(Type type, object? value, JsonSerializerOptions? options = null)
+            : this(type, value, CreateMediaType(JsonMediaType), options) { }
+
+        public JsonContent(Type type, object? value, string mediaType, JsonSerializerOptions? options = null)
+            : this(type, value, CreateMediaType(mediaType?? throw new ArgumentNullException(nameof(mediaType))), options) { }
+
+        public JsonContent(Type type, object? value, MediaTypeHeaderValue mediaType, JsonSerializerOptions? options = null)
+            : this(JsonSerializer.SerializeToUtf8Bytes(value, type, options), type, value, mediaType) { }
```

Comment:
- Can we serialize into a pooled `byte[]` somehow?
- We potentially end up buffering quite a lot here. Consider for future a way to "incrementally serialize up to X bytes", and if over X bytes, write to stream directly.
- I don't know the result of your recursive serialization support; is there any risk of running out of memory here?
- If a user passes `mediaType.CharSet` other than UTF-8, we'll be writing UTF-8 bytes but the other end will be decoding that as something else. We should either transcode here, or throw.

Reply:
I will switch the `JsonContent` implementation to use an underlying MemoryStream instead and will call `SerializeAsync`.

> Can we serialize into a pooled byte[] somehow?

`System.Text.Json` already implements a pooled `byte[]` buffer.

> If a user passes mediaType.CharSet other than UTF-8, we'll be writing UTF-8 bytes

I will validate by throwing on `CharSet`s != UTF-8 and then will add support for transcoding on both ends once https://github.com/dotnet/runtime/issues/30260 is done, unless this is something that can't wait; @terrajobst @mkArtakMSFT is this a blocker for Blazor?

Reply:
> I will switch the JsonContent implementation to use an underlying MemoryStream instead and will call SerializeAsync.

Why? It shouldn't need an intermediate stream. You can store the value, and then in SerializeToStreamAsync serialized directly to the destination stream.

> add support for transcoding on both ends once #30260 is done

The API proposed there is just a helper on top of existing APIs (with internals access it may also be able to optimize a bit further). You shouldn't be blocked by its absence, and even when it is available, it would only be for .NET 5, and this is building for other targets. We could of course decide the efficiency only matters for .NET 5, but I'll leave that to Immo on scenarios.

Reply:
> You can store the value, and then in SerializeToStreamAsync serialized directly to the destination stream.

That is true, just a side note; JsonSerializer validations (e.g: Type shouldn't be null or `Type.IsAssignableFrom(value.GetType())`) will occur until we send the content.

> it would only be for .NET 5, and this is building for other targets.

That is true as well, I will look for how other types handle this case.

Thanks for the suggestions.